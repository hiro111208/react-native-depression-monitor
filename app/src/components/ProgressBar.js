import React, { Component } from "react";
import { Platform } from "react-native";
import {
  Animated,
  StyleSheet,
  View,
  UIManager,
  LayoutAnimation,
  Text,
} from "react-native";
import BarMarkers from "./BarMarkers";

export default class ProgressBar extends Component {
  constructor() {
    super();
    this._animatedColor = new Animated.Value(0);
    // Enable LayoutAnimation under Android
    if (Platform.OS === "android") {
      UIManager.setLayoutAnimationEnabledExperimental(true);
    }
  }

  static defaultProps = {
    //width of current progress
    nextWidth: 0,
    segments: 4,
  };

  state = {
    //total width of progress bar container
    width: 0,
  };

  handleLayout = ({ nativeEvent }) => {
    this.setState({ width: nativeEvent.layout.width });
  };

  //calculate container width on mount
  componentDidMount() {
    this.handleLayout;
  }

  componentDidUpdate() {
    //custom Layout animation -> animates bar's width increase
    var CustomLayoutSpring = {
      duration: 1000,
      create: {
        type: LayoutAnimation.Types.spring,
        property: LayoutAnimation.Properties.scaleXY,
        springDamping: 0.7,
      },
      update: {
        type: LayoutAnimation.Types.spring,
        springDamping: 0.7,
      },
    };
    //set layout animation
    LayoutAnimation.configureNext(CustomLayoutSpring);

    //animate gradual color change
    this._animatedColor.setValue(0);
    Animated.timing(this._animatedColor, {
      toValue: 1,
      duration: 200,
      useNativeDriver: false,
    }).start();
  }

  render() {
    //constants to use
    const barSeparation = this.state.width / this.props.segments;
    const barWidthAfter = barSeparation * this.props.nextWidth;
    const colorChangeUnit =
      (255 / this.props.segments) * this.props.nextWidth - 1;
    const colorChangeUnitAfter =
      (255 / this.props.segments) * this.props.nextWidth;

    return (
      <View flexDirection={"row"} style={styles.shadow}>
        <View flexDirection={"column"} flex={1}>
          <View style={styles.barContainer} onLayout={this.handleLayout}>
            <Animated.View
              style={[
                styles.bar,
                {
                  width: barWidthAfter,
                  backgroundColor: this._animatedColor.interpolate({
                    inputRange: [0, 1],
                    outputRange: [
                      `rgb(${255 - colorChangeUnit},${0 + colorChangeUnit},0)`,
                      `rgb(${255 - colorChangeUnitAfter},${
                        0 + colorChangeUnitAfter
                      },0)`,
                    ],
                  }),
                },
              ]}
            >
              <View>
                <BarMarkers
                  bars={this.props.segments}
                  separation={barSeparation}
                ></BarMarkers>
              </View>
            </Animated.View>
          </View>
        </View>
        <View marginLeft={10}>
          <Text style={{ fontSize: 10, fontWeight: "300" }}>Progress</Text>
        </View>
        <View marginLeft={10} style={styles.progress} padding={0}>
          <View marginLeft={10} width={30}></View>
          <Text
            margin={5}
            style={{ height: 15, width: 30, fontSize: 11 }}
            adjustsFontSizeToFit={true}
          >
            {""}
            {Math.floor((barWidthAfter / this.state.width) * 100)}%
          </Text>
        </View>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  shadow: {
    shadowColor: "#000",
    shadowOpacity: 0.1,
    shadowRadius: 1,
  },
  barContainer: {
    opacity: 1,
    borderWidth: 1,
    backgroundColor: "white",
    height: 15,
    borderRadius: 7,
    overflow: "hidden",
  },
  bar: {
    opacity: 0.7,
    backgroundColor: "green",
    height: 20,
    borderRadius: 7,
    position: "absolute",
    left: 0,
    top: 0,
    bottom: 0,
    maxWidth: "100%",
  },
  progress: {
    opacity: 1,
    borderWidth: 1,
    textAlign: "center",
    alignItems: "center",
    padding: 3,
    height: 15,
    backgroundColor: "white",
    borderRadius: 20,
  },
});
