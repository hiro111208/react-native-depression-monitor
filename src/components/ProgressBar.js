import React, { Component } from "react";
import {
  Animated,
  StyleSheet,
  View,
  LayoutAnimation,
  Text,
} from "react-native";
import BarMarkers from "./BarMarkers";

export default class ProgressBar extends Component {
  constructor() {
    super();
    this._animatedColor = new Animated.Value(0);
  }

  static defaultProps = {
    //width of current progress
    nextWidth: 0,
    segments: 5,
  };

  state = {
    //total width of progress bar container
    width: 0,
  };

  handleLayout = ({ nativeEvent }) => {
    this.setState({ width: nativeEvent.layout.width });
  };

  //animate gradual color change
  animatedValue = () => {
    this._animatedColor.setValue(0);
    Animated.timing(this._animatedColor, {
      toValue: 1,
      duration: 200,
      useNativeDriver: false,
    }).start();
  };

  //calculate container width on mount
  componentDidMount() {
    this.handleLayout;
  }
  componentDidUpdate() {
    //custom Layout animation -> animates bar's width increase
    var CustomLayoutSpring = {
      duration: 1000,
      create: {
        type: LayoutAnimation.Types.spring,
        property: LayoutAnimation.Properties.scaleXY,
        springDamping: 0.7,
      },
      update: {
        type: LayoutAnimation.Types.spring,
        springDamping: 0.7,
      },
    };
    //set layout animation
    LayoutAnimation.configureNext(CustomLayoutSpring);
  }

  render() {
    //constants to use
    const barSeparation = this.state.width / this.props.segments;
    const barWidthAfter = barSeparation * this.props.nextWidth;

    //array of color interpolations: adjust to nÂº of Segments
    const colorGradient = [
      this._animatedColor.interpolate({
        inputRange: [0, 1],
        outputRange: ["rgb(255,0,0)", "rgb(255,50,0)"],
      }),
      this._animatedColor.interpolate({
        inputRange: [0, 1],
        outputRange: ["rgb(255,50,0)", "rgb(100,170,0)"],
      }),
      this._animatedColor.interpolate({
        inputRange: [0, 1],
        outputRange: ["rgb(100,170,0)", "rgb(0,170,0)"],
      }),
      this._animatedColor.interpolate({
        inputRange: [0, 1],
        outputRange: ["rgb(0,170,0)", "rgb(0,255,0)"],
      }),
    ];

    this.animatedValue();

    return (
      <View style={styles.barContainer} onLayout={this.handleLayout}>
        <Animated.View
          style={[
            styles.bar,
            {
              width: barWidthAfter,
              backgroundColor: colorGradient[this.props.nextWidth - 1],
            },
          ]}
        >
          <View>
            <BarMarkers
              bars={this.props.segments}
              separation={barSeparation}
            ></BarMarkers>
          </View>
        </Animated.View>
        <View justifyContent={"center"} alignItems={"center"}>
          <Text>{Math.floor((barWidthAfter / this.state.width) * 100)}%</Text>
        </View>
      </View>
    );
  }
}
const styles = StyleSheet.create({
  barContainer: {
    opacity: 1,
    borderWidth: 1,
    backgroundColor: "white",
    height: 20,
    borderRadius: 20,
    overflow: "hidden",
  },
  bar: {
    opacity: 0.7,
    backgroundColor: "green",
    height: 20,
    borderRadius: 20,
    position: "absolute",
    left: 0,
    top: 0,
    bottom: 0,
    maxWidth: "100%",
  },
});
