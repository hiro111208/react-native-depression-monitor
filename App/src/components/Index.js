import React, { Component } from "react";
import { View, StyleSheet, Button } from "react-native";
import ProgressBar from "./ProgressBar";

export default class Index extends Component {
  state = {
    currentWidth: -1,
    segments: 4,
  };

  incrementBar = () => {
    if (this.state.currentWidth < this.state.segments - 1) {
      this.setState({ currentWidth: ++this.state.currentWidth });
    }
    console.log(this.state.currentWidth + 1);
  };

  decrementBar = () => {
    if (this.state.currentWidth > -1) {
      this.setState({ currentWidth: --this.state.currentWidth });
    }
    console.log(this.state.currentWidth + 1);
  };

  render() {
    return (
      <View style={styles.container}>
        <Button
          className="increment"
          title=" + "
          onPress={this.incrementBar}
        ></Button>
        <ProgressBar
          segments={this.state.segments}
          nextWidth={this.state.currentWidth + 1}
        ></ProgressBar>
        <Button
          className="decrement"
          title=" - "
          onPress={this.decrementBar}
        ></Button>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    height: "100%",
    justifyContent: "center",
    backgroundColor: "#E6E6FA",
  },
});
