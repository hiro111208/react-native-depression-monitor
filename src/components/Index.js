import React, { Component } from "react";
import { View, StyleSheet, Button } from "react-native";
import ProgressBar from "./ProgressBar";

export default class Index extends Component {
  state = {
    steps: -1,
    segments: 3,
  };

  incrementBar = () => {
    if (this.state.steps < this.state.segments - 1) {
      this.setState({ steps: ++this.state.steps });
    }
    console.log(this.state.steps);
  };
  decrementBar = () => {
    if (this.state.steps > -1) {
      this.setState({ steps: --this.state.steps });
    }
    console.log(this.state.steps);
  };

  render() {
    return (
      <View style={styles.container}>
        <Button title=" + " onPress={this.incrementBar}></Button>
        <ProgressBar
          segments={this.state.segments}
          currentWidth={this.state.steps}
          nextWidth={this.state.steps + 1}
        ></ProgressBar>
        <Button title=" - " onPress={this.decrementBar}></Button>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    height: "100%",
    justifyContent: "center",
    backgroundColor: "#E6E6FA",
  },
});
