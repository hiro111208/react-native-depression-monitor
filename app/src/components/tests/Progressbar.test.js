import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import ProgressBar from "../ProgressBar";
import Index from "../Index";

describe("Testing ProgressBar.js", () => {

  it("ProgressBar renders correctly", () => {
    const wrapper = shallow(<Index />);
    expect(wrapper.find("ProgressBar")).toHaveLength(1);
  });

  it("ProgressBar renders correctly across screens", () => {
    const tree = renderer.create(<ProgressBar />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("ProgressBar renders with width progress of 0 ", () => {
    const wrapper = shallow(<Index />);
    expect(wrapper.find("ProgressBar").props().nextWidth).toBe(0);
  });
});