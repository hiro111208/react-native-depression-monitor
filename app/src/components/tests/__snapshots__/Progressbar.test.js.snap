// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ProgressBar.js ProgressBar renders correctly across screens 1`] = `
<View
  flexDirection="row"
  style={
    Object {
      "shadowColor": "#000",
      "shadowOpacity": 0.1,
      "shadowRadius": 1,
    }
  }
>
  <View
    flex={1}
    flexDirection="column"
  >
    <View
      onLayout={[Function]}
      style={
        Object {
          "backgroundColor": "white",
          "borderRadius": 7,
          "borderWidth": 1,
          "height": 15,
          "opacity": 1,
          "overflow": "hidden",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "rgba(255, 0, 0, 1)",
            "borderRadius": 7,
            "bottom": 0,
            "height": 20,
            "left": 0,
            "maxWidth": "100%",
            "opacity": 0.7,
            "position": "absolute",
            "top": 0,
            "width": 0,
          }
        }
      >
        <View>
          <View
            style={
              Object {
                "backgroundColor": "transparent",
                "flexDirection": "row",
                "height": "100%",
                "width": "100%",
              }
            }
          >
            <View
              left={0}
              style={
                Object {
                  "backgroundColor": "black",
                  "flexDirection": "column",
                  "opacity": 0.5,
                  "width": 1,
                }
              }
            />
            <View
              left={0}
              style={
                Object {
                  "backgroundColor": "black",
                  "flexDirection": "column",
                  "opacity": 0.5,
                  "width": 1,
                }
              }
            />
            <View
              left={0}
              style={
                Object {
                  "backgroundColor": "black",
                  "flexDirection": "column",
                  "opacity": 0.5,
                  "width": 1,
                }
              }
            />
            <View
              left={0}
              style={
                Object {
                  "backgroundColor": "black",
                  "flexDirection": "column",
                  "opacity": 0.5,
                  "width": 1,
                }
              }
            />
            <View
              left={0}
              style={
                Object {
                  "backgroundColor": "black",
                  "flexDirection": "column",
                  "opacity": 0.5,
                  "width": 1,
                }
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
  <View
    marginLeft={10}
  >
    <Text
      style={
        Object {
          "fontSize": 10,
          "fontWeight": "300",
        }
      }
    >
      Progress
    </Text>
  </View>
  <View
    marginLeft={10}
    padding={0}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "white",
        "borderRadius": 20,
        "borderWidth": 1,
        "height": 15,
        "opacity": 1,
        "padding": 3,
        "textAlign": "center",
      }
    }
  >
    <View
      marginLeft={10}
      width={30}
    />
    <Text
      adjustsFontSizeToFit={true}
      margin={5}
      style={
        Object {
          "fontSize": 11,
          "height": 15,
          "width": 30,
        }
      }
    >
      
      NaN
      %
    </Text>
  </View>
</View>
`;
